services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/dao/GreRangeRepository.java

```diff
@@ -0,0 +1,211 @@
+        }
+    }
+
+    /**
+     * Allocate a key from GreRange repository
+     * @param rangeId
+     * @return range key in db
+     * @throws Exception Db operation or key assignment exception
+     */
+    public synchronized Long allocateGreKey (String rangeId) throws Exception {
+        Long key;
+
+        try (Transaction tx = cache.getTransaction().start()) {
+            NetworkGRERange networkGRERange = cache.get(rangeId);
+            if (networkGRERange == null) {
+                throw new RangeNotFoundException();
+            }
+
+            key = networkGRERange.allocateKey();
+            cache.put(networkGRERange.getId(), networkGRERange);
+
+            tx.commit();
+        }
+
+        return key;
+    }
+
+    /**
+     * Release a key from GreRange repository
+     * @param rangId
+     * @param key
+     * @throws Exception Range Not Found Exception
+     */
+    public synchronized void releaseGreKey(String rangId, Long key) throws Exception {
+        try (Transaction tx = cache.getTransaction().start()) {
+            NetworkGRERange networkGRERange = cache.get(rangId);
+            if (networkGRERange == null) {
+                throw new RangeNotFoundException();
+            }
+
+            networkGRERange.release(key);
+            cache.put(networkGRERange.getId(), networkGRERange);
+
+            tx.commit();
+        }
+    }
+
+    public synchronized KeyAlloc getGreKeyAlloc(String rangeId, Long key) throws Exception {
+        NetworkGRERange networkGRERange = cache.get(rangeId);
+        if (networkGRERange == null) {
+            throw new RangeNotFoundException();
+        }
+
+        return networkGRERange.getNetworkKey(key);
+    }
+
+    /**
+     * Create a range entity from range repo
+     * @param request
+     * @throws Exception Network Range Already Exist Exception
+     * @throws Exception Internal Db Operation Exception
+     */
+    public synchronized void createRange(NetworkRangeRequest request) throws Exception {
+        try (Transaction tx = cache.getTransaction().start()) {
+            if (cache.get(request.getId()) != null) {
+                logger.warn("Create network range failed: Network Range already exists");
+                throw new NetworkRangeExistException();
+            }
+
+            NetworkGRERange range = new NetworkGRERange(request.getId(), request.getSegmentId(),
+                    request.getNetworkType(), request.getPartition(), request.getFirstKey(), request.getLastKey());
+
+            cache.put(request.getId(), range);
+
+            range = cache.get(request.getId());
+            if (range == null) {
+                logger.warn("Create network range failed: Internal db operation error");
+                throw new InternalDbOperationException();
+            }
+
+            request.setUsedKeys(range.getUsedKeys());
+            request.setTotalKeys(range.getTotalKeys());
+
+            tx.commit();
+        }
+
+    }
+
+    /**
```

Comment:
These comments are really good. Let us try to spread it elsewhere for important Controller and Service methods.

Reply:
Sure